File: Алгоритмы и структуры данных/Краскал/graph.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <math.h>

#include "unique_array.cpp"

using namespace std;

class graph {
private:
    bool is_letter(char a) {
        return (a >= 'a' && a <= 'z') || (a >= 'A' && a <= 'Z') || (a >= '0' && a <= '9') || a == '-';
    }

    bool is_tabulation(char a) {
        return a == '\t' || a == '\n';
    }

public:
    dynamic_array<edge> edges;
    unique_array<int, string> vertexes;

    graph(string path) {
        parsing_from_file(path);
    }

    void parsing_from_file(string path) {
        ifstream file(path);
        string line;
        getline(file, line);
        line += "\n";

        int cur_column = 0;
        int i = 0;
        while (i < line.size()) {
            string tmp;
            while (!is_tabulation(line[i])) {
                if (is_letter(line[i])) {
                    tmp += line[i];
                }
                i++;
            }
            i++;

            vertexes.put(cur_column, tmp);
            cur_column++;

            if (cur_column > 49) {
                throw new exception;
            }
        }

        for (int cur_line = 0; cur_line < vertexes.get_size(); cur_line++) {
            cur_column = 0;
            getline(file, line);
            line += "\n";

            i = 0;
            while (i < line.size()) {
                string tmp;
                while (!is_tabulation(line[i])) {
                    if (is_letter(line[i])) {
                        tmp += line[i];
                    }
                    i++;
                }
                i++;

                int weight = stoi(tmp);


                if (cur_line != cur_column && weight != 0) {
                    edge tmp(cur_line, cur_column, abs(weight));
                    edges.add(tmp);
                }
                cur_column++;
            }

            if (cur_column < vertexes.get_size() || cur_column > vertexes.get_size()) {
                throw new exception;
                break;
            }
        }
    }
};

File: Алгоритмы и структуры данных/Краскал на Java/Graph.java
package org.example;

import java.io.*;
import java.util.regex.*;

public class Graph
{
    public DynamicArray<Edge> edges = new DynamicArray<>();
    public UniqueArray<Integer, String> vertexes = new UniqueArray<>();

    public Graph(String path) throws IOException
    {
        parsingFromFile(path);
    }

    public void parsingFromFile(String path) throws IOException
    {
        BufferedReader graph = new BufferedReader(new FileReader(path));
        String line = graph.readLine() + '\n';

        Pattern pattern = Pattern.compile("[a-zA-Z0-9-]{1,5}[\t\n]");
        Matcher matcher = pattern.matcher(line);

        int curColumn = 0;
        while (matcher.find())
        {
            vertexes.put(curColumn, line.substring(matcher.start(), matcher.end() - 1));
            curColumn++;

            if (curColumn > 49)
            {
                throw new IndexOutOfBoundsException();
            }
        }

        for (int curLine = 0; curLine < vertexes.size(); curLine++)
        {
            curColumn = 0;
            line = graph.readLine() + "\n";

            matcher = pattern.matcher(line);

            while (matcher.find())
            {
                int weight = Integer.parseInt(line.substring(matcher.start(), matcher.end() - 1));
                if (curLine != curColumn && weight != 0)
                {
                    edges.add(new Edge(curLine, curColumn, Math.abs(weight)));
                }
                curColumn++;
            }

            if (curColumn < vertexes.size() || curColumn > vertexes.size())
            {
                throw new IndexOutOfBoundsException();
            }
        }
    }
}
